src/assignment-7/AMM.sol

```diff
@@ -0,0 +1,226 @@
+///     of the tokens reserves.
+/// @dev Explain to a developer any extra details
+contract AMM is ERC20Permit {
+
+    /******************
+     * Immutables
+     ******************/
+
+    /// @notice First ERC20 token of this AMM
+    IERC20 public immutable token0;
+
+    /// @notice Second ERC20 token of this AMM
+    IERC20 public immutable token1; 
+
+    /******************
+     * Storage
+     ******************/
+
+    /// @notice Amount of `token0` currently held by the contract
+    uint256 public reserve0;
+
+    /// @notice Amount of `token1` currently held by the contract
+    uint256 public reserve1;
+
+    /// @notice Indicates whether the contract has been initalized with initial reserves
+    bool public initalized;
+
+    /******************
+     * Events
+     ******************/
+
+    /// @notice Event emitted when first initalizing the AMM
+    event Initalized(uint256 amount0, uint256 amount1);
+
+    /// @notice Event emitted when a user deposits tokens into the AMM and gets LP tokens in return
+    event Mint(address indexed user, uint256 amount0, uint256 amount1, uint256 amountLP);
+
+    /// @notice Event emitted when a user burns their LP tokens in return for tokens from the reserve
+    event Burn(address indexed user, uint256 amount0, uint256 amount1, uint256 amountLP);
+
+    /// @notice Event emitted when a user sells `token0` into the AMM in return for `token1`
+    event Sell0(address indexed user, uint256 amount0, uint256 amount1);
+
+    /// @notice Event emitted when a user sells `token1` into the AMM in return for `token0`
+    event Sell1(address indexed user, uint256 amount0, uint256 amount1);
+
+    /******************
+     * Errors
+     ******************/
+
+    error AlreadyInitialized();
+    error IncorrectProportion();
+
+    /// @notice Creates a new AMM
+    /// @param token0_ First ERC20 token address of this AMM
+    /// @param token1_ Second ERC20 token address of this AMM
+    /// @param name The name of the AMM LP tokens
+    /// @param symbol The symbol of the AMM LP tokens
+    /// @param decimals The decimal numbers of the AMM LP tokens
+    constructor(
+        IERC20 token0_,
+        IERC20 token1_,
+        string memory name, 
+        string memory symbol, 
+        uint8 decimals) ERC20Permit(name, symbol, decimals) {
+
+        token0 = token0_;
+        token1 = token1_;
+    }
+
+    /// @notice Deposits the initial token reserves into the AMM
+    ///     The initalize exchange rate of the AMM will be set by the ratio between the token amounts
+    /// @dev This function can only be called once, will revert if called twice
+    /// @param amount0 Amount of `token0` to deposit
+    /// @param amount1 Amount of `token1` to deposit
+    /// @return lpTokens Amount of `lpTokens` minted to the user
+    function init(uint256 amount0, uint256 amount1) public returns (uint256 lpTokens) {
+        if (initalized) {
```

Comment:
You could also check that `totalSupply == 0`. That also solves the edge case where the initialized burns all his tokens before other LPs come along.

Reply:
👍  https://github.com/davidbrai/yield-mentorship/pull/8/commits/6e76ca18d9e935798ce411d65c47c28e580f5ba9